### ISSUE TITLE
Subcomponent rerender

### ISSUE THREAD

**Issue by mrgrimmig** [solkimicreb/react-easy-state#1]
Hello,

just found yesterday this amazing library (i've been following your projects nx-observe and nx-framework for some time).
So i wanted to check it out.

Now i have a small problem. I have to components: "parent" and "child".
The state of parent component looks like this: `this.state = { user: { name: "John Doe" } };`
I render the child component with user as a prop `<Child user={this.state.user} />`.

While changing the name in the parent or child component everything is fine `this.state.user.name = "foo";`. But when i change the complete user object it will propagate it to the child component but wouldn't rerender it `this.state.user = { name: "bar" };`. The parent component will be rerendered - the child component wouldn't.

Is it possible to do something like this?

**Comment by solkimicreb** [comment]
Everything should work just like with normal React. Could you answer these questions to find the issue quicker:

- Are both of your components class components?
- Do both of them have the `@easyState` decorator applied to them?
- Which version of Easy State are you using? (you should go with 1.03 (latest) as it has a critical fix).

This is a case when I let React do its own stuff (schedule its own render.) You change the `user` prop's reference, passed to the child component. In this case React should rerender the child by default. It might be a problem with Easy State's `shouldComponentUpdate` implementation, which tries to minimize rerenders. (I fixed an issue with it in 1.0.3) Could you try adding

```js
shouldComponentUpdate () {
  return true
}
```

to your child component?

Thx for the issue (:

**Comment by mrgrimmig** [comment]
Oh sorry, i discribed the problem a bit wrong.

1. `parent.state.user.name = "foo";` updates parent and child comp
2. `this.props.user.name = "baz";` will also update both components
3. `parent.state.user = { name: "bar" };` updates parent and child comp as well

but:
when i update the name (1 or 2) after updating the whole user object (3), something is broken.

example:
```js
// 1.)
parent.state.user.name = "foo" // "foo"
parent.state.user // Proxy {name: "foo"}
child.props.user // Proxy {name: "foo"}
// => parent and child displays "foo"

// 2.)
parent.state.user = { name: "baz" } // Object {name: "baz"}
parent.state.user //Proxy {name: "baz"}
child.props.user //Proxy {name: "baz"}
// => parent and child displays "baz"

// 3.)
parent.state.user.name = "john" // "john"
parent.state.user // Proxy {name: "john"}
child.props.user // Proxy {name: "john"}
// => parent displays "john", child still displays "baz"

// 4.)
child.props.name = "doe" // "doe"
parent.state.user // Proxy {name: "john"}
child.props.user // Proxy {name: "john"}
// => nothing happens. parent displays "john" child displays "baz"
```

I'm using version 1.0.3, both components are classes and uses @easyState decorator.
```
shouldComponentUpdate () {
  return true
}
```
didn't help

**Comment by solkimicreb** [comment]
Thanks, that was enough to find it :+1:. Its an issue with Easy State's `shouldComponentUpdate` implementation. There are actually two issues here:

- Easy State overrides user defined `shouldComponentUpdate`. Returning `true` from it would fix the issues, but Easy State overwrites your implementation so it won't even run. Prioritizing user defined `shouldComponentUpdate` will be a one liner fix.

- Easy State's `shouldComponent` update only allows React to schedule it's own update on shallow prop changes, which is not enough. I will check it and make it a bit more forgiving in the afternoon.

Thx for the help and detailed response (; Going the ship the fix later today.

**Comment by solkimicreb** [comment]
I fixed it and released v1.0.4. Change log is here: https://github.com/solkimicreb/react-easy-state/releases/tag/v1.0.4.

With this version the child comp should rerender and in case the parent comp only passes through the user and doesn't use the name it should not rerender on name changes (minimal rendering).

Let me know if it is still misbehaving on your side (:

**Comment by mrgrimmig** [comment]
Nice, the problem is fixed.
Thanks for the quick response.
